### PR TITLE
feat: Adding ephemeral disk support

### LINKED ISSUE #1287: Support ephemeral disks for agent nodes
**Is this an ISSUE or FEATURE REQUEST?** (choose one): Feature request

---

**What version of aks-engine?**: (all)

---

**Kubernetes version**: N/A

**What happened**:

Azure now offers [ephemeral OS disks](https://azure.microsoft.com/en-us/updates/announcing-ephemeral-os-disk-in-public-preview/) in public preview. These deploy and run faster because the IO remains local to the compute node. If the VM is moved, that data is lost. That's ok for agent nodes because persistent storage should be on Azure Disk or Azure Files, not the local nodes.

I did some manual tests of the new ephemeral disk option using Windows containers. This has some substantial performance improvements because it allows the VM to get better disk throughput, while taking advantage of the pre-pulled Windows Server Core images. 

VM Size tested: Standard_D2s_v3
VM Image: MicrosoftWindowsServer:WindowsServerSemiAnnual:Datacenter-Core-1809-with-Containers-smalldisk:1809.0.20190410
30GB OS disk

**Benchmark 1 - pull the lift & shift demo app**
              Measure-Command {docker pull patricklang/fabrikamfiber.web:1809 }

Without ephemeral disks: 6 mins, 56 seconds
With ephemeral disks: 4 mins, 58 seconds
**~28% reduction in pull time**

**Benchmark 2 - simple IO measurement on c:**
              diskspd.exe -c2G -w50 -b4K -r -o1 -W60 -d60 -Sh testfile.dat
              2G file, random 4k access, 50% read 50% write, single-threaded, 60s warmup, 60s testing, caching disabled

Without ephemeral disks: 120 IOPs
With ephemeral disks: 429 IOPs
**~3.5x faster**

**Anything else we need to know**:

/meow

### PR DESCRIPTION
**Pull request by PatrickLang**
**Reason for Change**:
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/ephemeral-os-disks make deployment and runtime faster.

This is still a WIP for review & testing. Will squash and make commit messages conventional soon.

**Issue Fixed**:
Fixes #1287

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [x] adds e2e test - will do on 7/25
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by PatrickLang** [comment]
This is ready for review.

**Comment by PatrickLang** [comment]
Manual tests looking good. Big improvement in 1st launch time of [iis](https://github.com/PatrickLang/Windows-K8s-Samples/blob/master/iis/iis-2019.yaml)

Clean deployment, nothing running

                        "imageReference": {
                            "publisher": "MicrosoftWindowsServer",
                            "offer": "WindowsServer",
                            "sku": "Datacenter-Core-1809-with-Containers-smalldisk",
                            "version": "17763.557.20190604"
                        }
OS disks were all left at 30Gb which is default for that image

2 Standard_D2s_v3 with ephemeral
pull succeeded in 4m 49s
ready in 4m 55s


2 Standard_D2s_v3 without ephemeral
pull succeeded in 6m 55s
ready in 7m 4s

**Comment by PatrickLang** [comment]
Nice, container local storage improved from ~100 IOPs to ~3300 IOPs

Results at https://github.com/PatrickLang/diskspd-containers

File: examples/disks-ephemeral/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# AKS Engine - Ephemeral Disks
+
+## Overview
+
+AKS Engine enables you to create customized Kubernetes cluster on Microsoft Azure with [ephemeral disks](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/ephemeral-os-disks).
+
+These examples are provided as a reference, note that managed disks is the default storage account type if none is specified.
```
Comment (jackfrancis) [inline examples/disks-ephemeral/README.md]: Should we add any copy that describes operational scenarios that would recommend an ephemeral disk configuration? Or do we need to do more discovery before publishing recommendations like that?

**Comment by jackfrancis** [comment]
lgtm, curious if we should merge this with more documentation language

**Comment by PatrickLang** [comment]
alright, will get more docs
/hold

**Comment by PatrickLang** [comment]
example warning:

```
$ ./bin/aks-engine generate /__w/1/s/gopath/src/github.com/Azure/aks-engine/_output/kubernetes-westeurope-63931.json --output-directory /__w/1/s/gopath/src/github.com/Azure/aks-engine/_output/kubernetes-westeurope-63931
time="2019-07-31T19:39:42Z" level=warning msg="Ephemeral disks are enabled for Agent Pool linuxpool1. This feature in AKS-Engine is experimental, and data could be lost in some cases."
time="2019-07-31T19:39:42Z" level=warning msg="Ephemeral disks are enabled for Agent Pool agentwfast. This feature in AKS-Engine is experimental, and data could be lost in some cases."
2019/07/31 19:39:42 
Kubeconfig:/__w/1/s/gopath/src/github.com/Azure/aks-engine/_output/kubernetes-westeurope-63931/kubeconfig/kubeconfig.westeurope.json
```

**Comment by PatrickLang** [comment]
Although many operations are faster, the E2E test pass duration impact is still within stddev

```
	PR#1651	PR#1700
1.12	0:38:47	0:50:55
1.13	0:44:33	0:39:00
1.14	0:37:10	0:44:55
1.15	0:36:57	0:39:51
1.16	0:51:19	0:49:36
mean	0:41:45	0:44:51
stddev	0:06:10	0:05:27
```

This PR leaves ephemeral on for the Windows test passes so we can measure if it breaks. If it's too brittle, then we can disable it.

File: examples/disks-ephemeral/README.md (hunk 1 above)
Comment (PatrickLang) [reply examples/disks-ephemeral/README.md]: I'm not ready to talk about specific workloads, but I did create a section on the features page for this that describes the state and risks.

**Comment by jackfrancis** [comment]
and

**Comment by palma21** [comment]
@PatrickLang do we have a sample model or doc for this? Wanted to take it for a spin

**Pull request merged**